### PR TITLE
sanity check for bad data

### PR DESCRIPTION
**Pull request by en0ma**
An optimistic approach to marking bad content, so that there are not sent to SPs (as they will fail during data transfer).

The first time `commp` or boost fails to pull blocks due to missing blocks, this will mark all contents affected by that block, so we stop sending deals for them and manually address such contents. This will reduce spamming SPs with deals that will fail due to bad content.

**verification**
I have tested it by trying to generate `commp`, trying the health check endpoint, and trying to access the content via gateway, in all 3 cases, the check worked!

- Estuary requesting commp from a shuttle for content with missing block, the shuttle sent back a sanity check (look at the time difference between the logs - its performant, cc @alvin-reyes ) to Estuary and Estuary marks the content.
Screenshot 2022-11-26 at 01 20 26

- Next, calling gateway and content health endpoint on the shuttle for same data, shuttle sends sanity check to Estuary to mark it
Screenshot 2022-11-26 at 01 20 43

- In all 3 cases, the sanity table only records 1 entry as expected
Screenshot 2022-11-26 at 01 45 17

- Lastly, trying to make deal, the content is ignored since it has missing blocks.
Screenshot 2022-11-26 at 01 30 53

-

### CODE REVIEW

**Comment by jcace** [comment]
nice! I like this abstraction

**Review by jlogelin — DISMISSED** [review]
LGTM

**Review by jcace — APPROVED** [review]

File: contentmgr/sanity_check.go [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+	if err := cm.DB.Find(&cnts, where, cc.Bytes()).Error; err != nil {
+		cm.log.Errorf("sanity check failed to get content(s) for cid: %s, err: %w", cc.String(), err)
+		return
+	}
+
+	// mark all contents affected by this missing block
+	var marks map[uint]bool
+	for _, cnt := range cnts {
+		if _, ok := marks[cnt.ID]; !ok {
+			if err := cm.DB.Model(&util.Content{}).Where("id = ?", cnt.ID).Update("failed_sanity_check", true).Error; err != nil {
```
Comment (alvin-reyes) [inline contentmgr/sanity_check.go:21]: Are we introducing a new field? I suggest you create a roll forward and back SQL script with transactions enabled.

File: contentmgr/shuttle.go [hunk 2]
```diff
@@ -220,6 +220,13 @@ func (cm *ContentManager) processShuttleMessage(handle string, msg *drpc.Message
 			cm.log.Errorf("handling split complete message from shuttle %s: %s", handle, err)
 		}
 		return nil
+	case drpc.OP_SanityCheck:
+		sc := msg.Params.SanityCheck
+		if sc == nil {
+			return ErrNilParams
+		}
+		cm.HandleSanityCheck(sc.CID, sc.Err)
```
Comment (alvin-reyes) [inline contentmgr/shuttle.go:228]: I think this is going to be an expensive operation (it looks up obj and obj ref tables). I’m thinking if we should decouple this as part of the GC.

File: util/content.go [hunk 3]
```diff
@@ -99,6 +99,8 @@ type Content struct {
 	// them (unlike with aggregates)
 	DagSplit  bool `json:"dagSplit"`
 	SplitFrom uint `json:"splitFrom"`
+
+	FailedSanityCheck bool `json:"-"`
```
Comment (alvin-reyes) [inline util/content.go:103]: Introducing a new column in a massive table like Content needs to have a roll forward and back script. We will need to execute this SQL script on API node first before we deploy the app.

File: contentmgr/shuttle.go (hunk 2 above)
Comment (en0ma) [reply contentmgr/shuttle.go:228]: I do not think so, we do this all the time when we do commp, plus missing blocks do not happen often (and should not)

File: contentmgr/sanity_check.go (hunk 1 above)
Comment (alvin-reyes) [reply contentmgr/sanity_check.go:21]: I'll add this to the instructions for the next release.

file_name: contents_add_failed_sanity_check_column_roll_forward.sql
```
\echo 'start - adding failed_sanity_check on contents table'
// Roll forward
alter table contents add failed_sanity_check boolean;
\echo 'end - adding failed_sanity_check on contents table'
```

file_name: contents_add_failed_sanity_check_column_roll_back.sql
```
\echo 'start - adding failed_sanity_check on contents table'
alter table contents drop column failed_sanity_check;
\echo 'end - adding failed_sanity_check on contents table'
```

Comment (en0ma) [reply contentmgr/sanity_check.go]: @alvin-reyes I think it best to use another table, so we can persist the missing block and the error message - which might come in handy for manually fixing the data, plus this won't strain the contents table

File: util/content.go (hunk 3 above)
Comment (en0ma) [reply util/content.go]: I have changed it to use a new table

**Comment by en0ma** [comment]
@jcace @jlogelin @alvin-reyes I have tested this with one of the bad data - it works!

**Review by jcace — APPROVED** [review]
Looks good!

**Review by alvin-reyes — APPROVED** [review]
LGTM

**Comment by alvin-reyes** [comment]
Great work on this. Let's get this included on v0.2.0.

**Pull request merged**